Fortunac/fun specs flag

fortunac opened the pull request:
Adds a flag so the user can specify which built-in function specs they want to use. The built-in summaries are:
- verifier-error
- verifier-assume
- verifier-nondet
- afl-maybe-log
- arg-terms
- chaos-caller-saved
- rax-out
- chaos-rax

Also checks which subroutines are to be inlined. If a subroutine is given both the inline spec and a summary spec, this will report an error rather than picking one spec over the other.

codyroux commented on wp/plugin/lib/analysis.ml:
```diff
@@ -124,20 +124,10 @@ let exp_conds_mod (p : Params.t) : Env.exp_cond list =
 (* Determine which function specs to use in WP. *)
 let fun_specs (p : Params.t) (to_inline : Sub.t Seq.t)
```
Per our Wednesday conversation: what is the default behavior if the user does not set the flag?

fortunac replied:
The current default behavior is to set the function as called and update the stack pointer. I plan on changing this to chaosing the caller-saved registers, and if a function is one of the special functions from our verifier api, use those corresponding specs.

codyroux replied:
👍 just make sure this appears in the comments/docs somewhere.

codyroux commented on wp/resources/sample_binaries/nested_ifs/run_wp_inline.sh:
```diff
@@ -8,7 +8,7 @@
   bap wp \
     --func=main \
-    --inline=.* \
+    --inline=calloc \
```
Is this a fix for an older issue?

fortunac replied:
Since I removed VERIFIER_nondet from the default list of function specs, inlining everything, would inline calloc, and that would make this test fail. When I add the nondet spec back to the default, I think I could change this back to `.*`.

codyroux replied:
👍

codyroux commented on wp/plugin/wp.ml:
```diff
@@ -207,6 +207,36 @@ let func_name_map = Cmd.parameter Typ.(list ~sep:';' (pair ~sep:',' string strin
+let fun_specs = Cmd.parameter Typ.(list string) "fun-specs"
+    ~doc:{|List of built-in function summaries to be used at a function call
+           site in order of precedence. A target function will be mapped to a
+           function spec if it fulfills the spec's requirements. All function
```
Should we say more about the "spec's requirements"?

fortunac replied:
Can do!

codyroux commented on wp/plugin/wp.ml:
```diff
@@ -207,6 +207,36 @@ let func_name_map = Cmd.parameter Typ.(list ~sep:';' (pair ~sep:',' string strin
+           pointer. The default spec will set the function as called and update
+           the stack pointer. Available built-in specs:
+
+           `verifier-error': Trips calls to __assert_fail.
```
What is "Trips calls"?

fortunac replied:
It would get WP to look for inputs to the subroutine that would cause us to reach `__assert_fail' or `__VERIFIER_error'. I can add this.

codyroux replied:
I just want the docs to explain what "trips calls" mean to a casual user.

codyroux commented on wp/plugin/wp.ml:
```diff
@@ -207,6 +207,36 @@ let func_name_map = Cmd.parameter Typ.(list ~sep:';' (pair ~sep:',' string strin
+           `verifier-assume': Adds an assumption to the precondition based on
+           the argument to the function call.
+
+           `verifier-nondet': Chaoses the output to the function call
```
It feels like these correspond to specific functions.

fortunac replied:
I'll add this too.

codyroux replied:
👍

fortunac commented:
Changed the default specs to be `verifier-assume`, `verifier-nondet`, `chaos-caller-saved`, and `empty`. I did not add `verifier-error` to the default list of specs since it is controlled by the `trip-asserts` flag (although the user can stilll specify it with the `fun-specs` flag). I also updated the docs to include the list of default specs and the functions each spec corresponds to.
